### ISSUE TITLE
twisted.conch.telnet.Telnet.requestNegotiation sends the wrong bytes for the requested negotiation

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3655]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3655|
|Type|defect|
|Created|2009-02-12 21:59:21Z|
|Branch|https://github.com/twisted/twisted/tree/telnet-subnegotiation-3655|

```python
    def requestNegotiation(self, about, bytes):
        bytes = bytes.replace(IAC, IAC * 2)
        self._write(IAC + SB + bytes + IAC + SE)
```

The disuse of the `about` parameter is a hint about how the method is broken (it should be sending its value as part of that string it writes).

This prevents any outgoing subnegotiation initiated by `Telnet` from working properly.

<details><summary>Searchable metadata</summary>

```
trac-id__3655 3655
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_telnet_subnegotiation_3655 branches/telnet-subnegotiation-3655
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1234475961000000 1234475961000000
changetime__1234668548000000 1234668548000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [26215]) Branching to 'telnet-subnegotiation-3655'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/telnet-subnegotiation-3655)

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

The change looks correct, but I was unable to see how it changes the behavior. Is there a simple way of seeing it by running a Twisted telnet server and connect netkit telnet client to it?

That's a blocker point though, please merge, after updating the copyright of telnet.py.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> The change looks correct, but I was unable to see how it changes the behavior. Is there a simple way of seeing it by running a Twisted telnet server and connect netkit telnet client to it? 

There's not actually a telnet server in Twisted that uses this feature. :/  Hm, let's see...

```python
import sys

from twisted.python import log
from twisted.conch.telnet import TelnetTransport, TelnetProtocol
from twisted.internet.protocol import ServerFactory
from twisted.internet import reactor

TTYPE = chr(24)
SEND = chr(1)

class TerminalTypeReporter(TelnetProtocol):
    def connectionMade(self):
        TelnetProtocol.connectionMade(self)
        self.transport.negotiationMap[TTYPE] = self._ttype
        self.transport.do(TTYPE).addCallbacks(
            lambda ignored:
                self.transport.requestNegotiation(TTYPE, SEND),
            log.err)


    def enableRemote(self, option):
        return option == TTYPE

    def _ttype(self, bytes):
        print 'Terminal Type:', bytes      
        self.transport.loseConnection()


factory = ServerFactory()
factory.protocol = lambda: TelnetTransport(TerminalTypeReporter)
reactor.listenTCP(12345, factory)

log.startLogging(sys.stdout)
reactor.run()
```

should demonstrate it.  Without the fix, connecting to the server will do nothing.  With the fix, the client will be disconnected immediately and the server will log a message about the kind of terminal the client is using (assuming the client supports this option - but netkit does).

> That's a blocker point though, please merge, after updating the copyright of telnet.py. 

I **think** you meant to say _not_ in there.  If not, then I'll try to fix it later. :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [26225]) Merge telnet-subnegotiation-3655

Author: exarkun
Reviewer: therve
Fixes: [#3655](https://github.com/twisted/twisted/issues/3655)

Change `twisted.conch.telnet.Telnet.requestNegotiation` so that it generates the correct string
to send to the peer to negotiate about the specified option.  Previously the implementation was
buggy in that it omitted the option byte, preventing a subnegotiation request from ever completing
as intended.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
